src/lem/constrainer.rs

```diff
@@ -81,145 +117,320 @@ impl LEM {
         Ok(())
     }
 
-    fn on_concrete_path(concrete_path: &Option<Boolean>) -> Result<bool> {
-        match concrete_path {
-            None => Ok(true),
-            Some(concrete_path) => concrete_path
-                .get_value()
-                .ok_or_else(|| anyhow!("Couldn't check whether we're on a concrete path")),
-        }
-    }
-
     fn allocate_and_inputize_input<'a, F: LurkField, CS: ConstraintSystem<F>>(
         &'a self,
         cs: &mut CS,
         store: &mut Store<F>,
-        witness: &Witness<F>,
+        valuation: &Valuation<F>,
         alloc_ptrs: &mut HashMap<&'a String, AllocatedPtr<F>>,
         idx: usize,
     ) -> Result<()> {
-        if alloc_ptrs.contains_key(&self.input[idx]) {
-            bail!("{} already allocated", &self.input[idx]);
-        }
         let alloc_ptr = Self::allocate_ptr(
-            &mut cs.namespace(|| format!("allocate input {}", &self.input[idx])),
-            &store.hydrate_ptr(&witness.input[idx])?,
+            cs,
+            &store.hydrate_ptr(&valuation.input[idx])?,
             &self.input[idx],
+            alloc_ptrs,
         )?;
         alloc_ptrs.insert(&self.input[idx], alloc_ptr.clone());
         Self::inputize_ptr(cs, &alloc_ptr, &self.input[idx])?;
         Ok(())
     }
 
-    /// Create R1CS constraints for LEM given an evaluation witness.
+    fn on_concrete_path(concrete_path: &Boolean) -> Result<bool> {
+        concrete_path
+            .get_value()
+            .ok_or_else(|| anyhow!("Couldn't check whether we're on a concrete path"))
+    }
+
+    fn z_ptr_from_valuation<F: LurkField>(
+        concrete_path: &Boolean,
+        valuation: &Valuation<F>,
+        name: &String,
+        store: &mut Store<F>,
+    ) -> Result<ZPtr<F>> {
+        if Self::on_concrete_path(concrete_path)? {
+            let Some(ptr) = valuation.ptrs.get(name) else {
+                bail!("Couldn't retrieve {} from valuation", name);
+            };
+            store.hydrate_ptr(ptr)
+        } else {
+            Ok(ZPtr::dummy())
+        }
+    }
+
+    /// Accumulates slot data that will be used later to generate the constraints.
+    fn acc_slots_data_data<F: LurkField>(
+        concrete_path: Boolean,
+        slots_data: &mut SlotsData<F>,
+        img: AllocatedNum<F>,
+        img_name: String,
+        preimg_vec: Vec<&AllocatedPtr<F>>,
+    ) {
+        let arity = match preimg_vec.len() {
+            2 => HashArity::A2,
+            3 => HashArity::A3,
+            4 => HashArity::A4,
+            _ => unreachable!(),
+        };
```

Comment:
This should eventually get factored out to a method of `HashArity`, I think. Especially since you have a default match arm here, changes to `HashArity` won't trigger compiler errors. So one will have to manually find and update all such examples scattered through the code. You could minimize by making this more clearly part of the `HashArity` implementation — maybe by implementing `From<usize>` for `HashArity`. Also, I think this is only unreachable if it's impossible for a preimage of an unsupported size to be passed in here. Is that the case, or could a 'bad LEM program' do so?

Reply:
I agree. And it's truly unreachable because no `LEMOP` has arities out of that range